File: app/src/main/java/pe/andy/bookholic/ui/ScrollToTopButton.java
package pe.andy.bookholic.ui;

import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.ImageView;

import androidx.core.widget.NestedScrollView;

import pe.andy.bookholic.MainActivity;
import pe.andy.bookholic.databinding.MainActivityBinding;

public class ScrollToTopButton {

    MainActivityBinding mBinding;
    ImageView gotoTop;

    public ScrollToTopButton(MainActivity activity) {
        this.mBinding = activity.getMBinding();

        this.gotoTop = this.mBinding.goToTop;
        this.gotoTop.setVisibility(View.GONE);

        this.gotoTop.setOnClickListener(view -> {
            this.mBinding.nestedScrollView.smoothScrollTo(0, 0);
        });

        this.mBinding.nestedScrollView.setOnScrollChangeListener(
            (NestedScrollView v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) -> {

                final int UP = -1;
                final int DOWN = 1;

                boolean canScrollUp = v.canScrollVertically(UP);
                boolean canScrollDown = v.canScrollVertically(DOWN);

                // ON TOP
                if ((!canScrollUp) && canScrollDown) {
                    hide();
                }
                // ON BOTTOM
                else if (canScrollUp && (!canScrollDown)) {
                    hide();
                }
                else if (this.mBinding.goToTop.getVisibility() == View.GONE) {
                    show();
                }
        });
    }

    void show() {
        Animation animation = new AlphaAnimation(0, 1);
        animation.setDuration(500);
        this.mBinding.goToTop.setVisibility(View.VISIBLE);
        this.mBinding.goToTop.setAnimation(animation);
    }

    void hide() {
        Animation animation = new AlphaAnimation(1, 0);
        animation.setDuration(500);
        this.mBinding.goToTop.setVisibility(View.GONE);
        this.mBinding.goToTop.setAnimation(animation);
    }

}

File: app/src/main/java/pe/andy/bookholic/ui/ScrollToTopButton.kt
package pe.andy.bookholic.ui

import android.view.View
import android.view.animation.AlphaAnimation
import androidx.core.widget.NestedScrollView
import pe.andy.bookholic.databinding.MainActivityBinding

class ScrollToTopButton(private var mBinding: MainActivityBinding) {

    private val showAnimation = AlphaAnimation(0f, 1f)
    private val hideAnimation = AlphaAnimation(1f, 0f)

    init {
        showAnimation.duration = 500
        hideAnimation.duration = 500

        mBinding.goToTop.visibility = View.GONE
        mBinding.goToTop.setOnClickListener {
            mBinding.nestedScrollView.smoothScrollTo(0, 0)
        }

        mBinding.nestedScrollView.setOnScrollChangeListener {
            v: NestedScrollView?, _: Int, _: Int, _: Int, _: Int ->

            val UP = -1
            val DOWN = 1

            val canScrollUp = v?.canScrollVertically(UP) ?: false
            val canScrollDown = v?.canScrollVertically(DOWN) ?: false

            // Top or Bottom
            if (canScrollUp xor canScrollDown) {
                hide()
            }
            // Middle
            else if (mBinding.goToTop.visibility == View.GONE) {
                show()
            }

        }
    }

    private fun show() {
        mBinding.goToTop.visibility = View.VISIBLE
        mBinding.goToTop.animation = showAnimation
    }

    private fun hide() {
        mBinding.goToTop.visibility = View.GONE
        mBinding.goToTop.animation = hideAnimation
    }

}